x/delayedack/keeper/rollapp_packet.go

```diff
@@ -120,61 +124,33 @@ func (k *Keeper) UpdateRollappPacketWithStatus(ctx sdk.Context, rollappPacket co
 	return rollappPacket, nil
 }
 
-// ListRollappPacketsByStatus retrieves a list of pending rollapp packets from the KVStore.
-// It builds a prefix using the rollappID and the pending status, and iterates over the range from lastProofHeight to proofHeight.
-// If the packet's proofHeight is less than or equal to the maxProofHeight, it is added to the list.
-// if maxProofHeight is 0, all packets are returned.
-// The function returns the list of pending packets.
-func (k Keeper) ListRollappPacketsByStatus(
-	ctx sdk.Context,
-	status commontypes.Status,
-	maxProofHeight uint64,
-) (list []commontypes.RollappPacket) {
+// ListRollappPackets retrieves a list of rollapp packets from the KVStore by applying the provided filter
+func (k Keeper) ListRollappPackets(ctx sdk.Context, listFilter rollappPacketListFilter) (list []commontypes.RollappPacket) {
 	store := ctx.KVStore(k.storeKey)
-	// switch prefix based on status
-	statusPrefix := commontypes.MustGetStatusBytes(status)
-	// Iterate over the range from lastProofHeight to proofHeight.
-	// We are guaranteed order by the proof height so can break early if we
-	// find a packet with a proof height greater than maxProofHeight
-	iterator := sdk.KVStorePrefixIterator(store, statusPrefix)
+	iterator := sdk.KVStorePrefixIterator(store, listFilter.prefixBytes)
+
 	defer iterator.Close() // nolint: errcheck
 
 	for ; iterator.Valid(); iterator.Next() {
 		var val commontypes.RollappPacket
 		k.cdc.MustUnmarshal(iterator.Value(), &val)
-		if maxProofHeight == 0 || val.ProofHeight <= maxProofHeight {
+
+		if listFilter.filter == nil || listFilter.filter(val) {
 			list = append(list, val)
-		} else {
+		} else if listFilter.stopOnFirstMatch {
```

Comment:
imo `stopOnFirstMatch` is a confusing name as it doesn’t actually stop on first match but on first "Unmatch" in that sense.

Reply:
my bad